### DISCUSSION TITLE
Significant delay when redirecting to getStaticProps() page as it waits for data fetching. No `router.isFallback` loading state is shown

### DISCUSSION THREAD

**Discussion by AshConnolly** [vercel/next.js#16254] [Help]
# Bug report

## Describe the bug

When clicking a link that redirects to a page that uses getStaticProps() there is a significant delay before the user see's any changes. This is because it's waiting for all the data fetching in the target page's getStaticProps() to finish downloading, before redirecting the user. 

If a user visits a getStaticProps() page directly however, they will see the loading state of:
```
  if (router.isFallback) {
    return <div>Loading...</div>
  }
```

## To Reproduce

1. Go to [codesandbox](https://codesandbox.io/s/nextjs-redirect-to-getstaticprops-page-issue-significant-delay-xtwvx?file=/pages/index.js)
2. click the links and see the above behaviour.

## Expected behavior

When a user clicks a link they should get immediate feedback in being be sent to the destination page (url change) and instantly see a loading state while it fetches data. This would also make the behavior consistent, as currently we only see loading on a direct visit (no redirect) to a `getstaticprops()` page. 

## Additional Comments

I think the expected behavior outlined above is the default behavior any user would expect. It's a confusing UX to click a link and wait for the resources to be retrieved before giving a user any feedback. Perhaps this _"wait until assets are fetched before redirecting"_ could be a developer option like 
```
<Link gspWait href={\`/github/[user]`} as={`/github/timneutkens`}>
```

Currently in my app I have had to do full redirects with standard anchor tags to get around this, which has a lot of downsides, both for performance and losing app state.

**Comment by timneutkens** [comment]
It's expected behavior that link navigations are blocking in this way currently, initially we had added the behavior that you're describing but it added some very impractical issues where you see flashes of loading states when the data loads right after rendering the loading state and such. Suspense makes this a bit better but that's not available yet in mainstream React.

**Reply by AshConnolly** [reply]
Hi Tim! 👋  Ah right, thanks for the explanation. Yeh I can see how Suspense would help.

I still wonder if it's potentially a more confusing issue for users to not see anything happen on screen while they wait. I don't how other frameworks do it but seems odd to click and not see something happen instantly, this will have a negative impact on the [perceived performance](https://www.smashingmagazine.com/2015/09/why-performance-matters-the-perception-of-time/) (and the UX) of an app.

Could we not make this behavior optional? Like we could choose to immediately send them and risk the flash of loading state? Because the amount of data being fetched from the getStaticProps() and time to download it is very app specific. 

**Do you have any recommended solutions for this issue**, as I mentioned above I have had to do full redirects with standard anchor tags, which has a lot of performance and state downsides. Is there a way to show the user something during this wait?

**Reply by timneutkens** [reply]
You can add a loading bar to your application: https://github.com/vercel/next.js/tree/canary/examples/with-loading

> Could we not make this behavior optional? Like we could choose to immediately send them and risk the flash of loading state? Because the amount of data being fetched from the getStaticProps() and time to download it is very app specific.

If opt-in it's unlikely it'll get adoption anywhere near what we'd want to see for something like this, hence why it's not in the framework currently.

**Reply by AshConnolly** [reply]
(Late reply, sry!) Thanks for that Tim! Appreciate it! 👍

**Reply by arklanq** [reply]
It really should be described in documentation of next. Right now informations about `fallback: true` option in docs are totally missleading, becuase it present different behaviour from what I have seen in docs and what i have encountered my-self in my app.

https://nextjs.org/docs/basic-features/data-fetching#fallback-true
